### ISSUE TITLE
Bash formatting not working properly (sometimes)

### ISSUE THREAD

**Issue by shinebayar-g** [bugy/script-server#169]
I have this simple script that pull new changes from repo and build docker image & push.

```shell
#!/bin/bash
# simple docker build script

now=$(date +'%H:%M:%S')
echo -e "\e[1mStep1 $now: Changing directory to dependencies\e[0m"
cd /home/shinee/web/service/dependencies-base

now=$(date +'%H:%M:%S')
echo -e "\e[1mStep2 $now: Pulling update from master\e[0m"
git pull origin master

now=$(date +'%H:%M:%S')
echo -e "\e[1mStep3 $now: Updating image node:8-slim\e[0m"
docker pull node:8-slim

now=$(date +'%H:%M:%S')
echo -e "\e[1mStep4 $now: Updating image nginx:stable\e[0m"
docker pull nginx:stable

now=$(date +'%H:%M:%S')
echo -e "\e[1mStep5 $now: Building image myrepo/image:tag\e[0m"
#docker build --force-rm -t myrepo/image:tag -f Dockerfile . || { echo "couldn't build docker image" ; exit 1; }

now=$(date +'%H:%M:%S')
echo -e "\e[1mStep6 $now: Pushing image myrepo/image:tag\e[0m"
#docker push myrepo/image:tag
```

I intentionally commented out docker build & push part for the sake of testing. But bug is still reproducable.

To reproduce:
If you run this script from command line. `./test_build.sh` only line that contains `echo -e` bash formatted lines will be **bold**, and commands outputs will be normal text.
![image](https://user-images.githubusercontent.com/3091558/51583809-42f68980-1f0d-11e9-94d6-afb1476d4d0b.png)

But this same script's output doesn't look as expected in script-server window. As you see
```shell
echo -e "\e[1mStep2 $now : Pulling update from master\e[0m"    # this part should be formatted only
git pull origin master     # but this commands output are also getting formatted, sometimes. (most of the time)
```

![image](https://user-images.githubusercontent.com/3091558/51583885-a4b6f380-1f0d-11e9-924c-02912c3ce6cd.png)

**Comment by bugy** [comment]
Thanks for reporting

**Comment by bugy** [comment]
Hi @shinebayar-g , unfortunately I couldn't reproduce. I copied your script, replacing docker commands with simple echo. And `git pull`ing from another repo. But after 88 tries I couldn't achieve wrong formatting.
I also double-checked the code for any obvious errors and couldn't notice any.

How often does it happen to you? Do you by any chance have logs for this execution? Are they different from other logs?

**Comment by shinebayar-g** [comment]
Ok this bug appeared. When I `cat log` files, outputs looks as expected. How can I copy html code? I cannot get source by `view-source` function?

This is best I can do for now
![image](https://user-images.githubusercontent.com/3091558/51835772-6d937880-2339-11e9-9820-f1c3b1adf976.png)
As you see whole div gets **bold style**.
I also got this.
![image](https://user-images.githubusercontent.com/3091558/51835838-9287eb80-2339-11e9-91e5-5e9efea8f3d3.png)

**Comment by bugy** [comment]
Hi @shinebayar-g , thanks for providing all this info! 
Regarding HTML code: the screenshot is enough for me. And according to it, it's not a markup issue (but still could be UI or server-side)
The warnings shouldn't happen in a normal scenario and could be either effect of wrong text formatting or the cause for it.

Unfortunately this information is not enough for me, to make any other guesses/assumptions :( How often it reproduces for you? May be you could provide me a log file, which you `cat`, probably cut and obfuscated (and via email, to avoid publishing for everyone)

**Comment by bugy** [comment]
2 points:
1. May be you can try to track websocket data, and send it to me (it's available in dev tools, on network tab. And you need to open dev tools before starting the script)
2. I'm not sure if it makes sense to investigate the issue, it will be gone (or have different effect) after #174

**Comment by shinebayar-g** [comment]
Yea, let's forget about this issue until #174 is come to play.

**Comment by bugy** [comment]
So, #174 is there. Could you try again please, with the dev build, if possible? Actually you can even test it with jsut an admin panel, since it's displaying formatted output now

**Comment by shinebayar-g** [comment]
Hi @bugy Just updated my script-server with **dev** build.
Looks like bash formatting issue is fixed. Also I've checked my old logs, all looks liked fixed. 
However I've got another issue:
- When I run script it finishes like this. Looks like bottom messages are cut or something. 
Note: I checked script successfully run as it's updated docker image.
![image](https://user-images.githubusercontent.com/3091558/55286014-86b3b680-53c8-11e9-8177-79bc3a360746.png)
- If I check this script's log from admin menu, you'll notice script was finished successfully. 
![image](https://user-images.githubusercontent.com/3091558/55286044-ec07a780-53c8-11e9-805b-a9caa05c1eec.png)

IMO, there are 2 issues here.
1. Docker build command's logs aren't showing correctly. However end result looks correct from admin page.
2. Idk it's because of docker's output or not, my last bash command didn't show in script-server.

```
now=$(date +'%H:%M:%S')
echo -e "\e[1m$now : Update successfully!!\e[0m"
```

**Comment by bugy** [comment]
Hi @shinebayar-g, thanks for checking and reporting! If you see correct log in the admin page, then it's definitely Script server issue (not because of Docker): I write to the log file the same text as to UI. 
I see 2 potential causes here:
1. An error during sending data or rendering it in the browser
2. Race condition on finishing script and closing client connection.

The first one is easier to check: it should be logged either in server logs, or you would see some errors in the browser console. Could you check please, if you have any (especially in browser, when issue happens again).
To check the second one, it would be really nice of you, if you could run a script with developer tools open. And then in network tab check the data chunks, sent via socket.

How easy is it for you to reproduce it? Or it happens regularly?

**Comment by shinebayar-g** [comment]
Another one: My script's output messed up like this, it didn't really create new lines, but rewrite current lines..

![image](https://user-images.githubusercontent.com/3091558/55286216-40ac2200-53cb-11e9-8ba2-87510acab8b7.png)

**Comment by bugy** [comment]
Hi @shinebayar-g, I think there are some ANSI codes, which are not supported yet: #134. I support some cursor move commands (up/left/down/right), but not others (move to position x,y, move to the beginning, move to remembered position, etc.). I think Docker is mixing them and it causes troubles for you.
I was going to work on extending ANSI next week.

**Comment by shinebayar-g** [comment]
> Another one: My script's output messed up like this, it didn't really create new lines, but rewrite current lines..
> 
> ![image](https://user-images.githubusercontent.com/3091558/55286216-40ac2200-53cb-11e9-8ba2-87510acab8b7.png)

I'm seeing correct output on admin page. @bugy

**Comment by bugy** [comment]
Hi @shinebayar-g, interesting, thanks. Then it brings me to another assumption, which relates to internal implementation. And if it's true, then it could also affect the cut log.

**Comment by shinebayar-g** [comment]
Docker's output is working as expected on most of times. I don't know why it's bugged sometimes.. I'll always check developer console.. About network tab I'm not sure what to do?

![image](https://user-images.githubusercontent.com/3091558/55287136-98e92100-53d7-11e9-90fa-34fb28f3b0df.png)

**Comment by bugy** [comment]
Hi @shinebayar-g , please find the screenshot of network tab attached. On the screenshot I selected the socket connection (/exections/io), and on the right side "Frames" tab. This tab includes all the input/output data of a running script.
![image](https://user-images.githubusercontent.com/1275813/55287435-0df13200-53a9-11e9-9067-5ff5d4d14400.png)

**Comment by bugy** [comment]
Hi @shinebayar-g, I found one bug in algorithm, related to "move cursor up" command. The fix is committed and available in a dev build

PS your output file was a great help for me. I just `cat` it (in chunks) and reproduced the problem, thanks a lot! I also noticed some formatting issues with docker download output. May be another bug, may be unsupported ANSI characters. Will have a look

**Comment by shinebayar-g** [comment]
Just tested out new build. Run few scripts, didn't notice serious bug yet.

> I also noticed some formatting issues with docker download output. May be another bug, may be unsupported ANSI characters. Will have a look

I see, hope you'll catch it. 💯

**Comment by bugy** [comment]
So, there were both: missing ANSI command support (clear line) + a bug when moving cursor bottom. Fixed both.

Now Script server log looks exactly like terminal output. At least for the log file you sent me :)
